src/Compilers/CSharp/Portable/Binder/Binder_Patterns.cs

```diff
@@ -569,55 +605,161 @@ private BoundPattern BindRecursivePattern(RecursivePatternSyntax node, TypeSymbo
-                        if (parameterIndex < deconstructMethod.ParameterCount)
+                        parameter = deconstructMethod.Parameters[parameterIndex];
+                        string parameterName = parameter.Name;
+                        if (name != parameterName)
                         {
-                            parameter = deconstructMethod.Parameters[parameterIndex];
-                            string parameterName = parameter.Name;
-                            if (name != parameterName)
-                            {
-                                diagnostics.Add(ErrorCode.ERR_DeconstructParameterNameMismatch, subPattern.NameColon.Name.Location, name, parameterName);
-                            }
+                            diagnostics.Add(ErrorCode.ERR_DeconstructParameterNameMismatch, subPattern.NameColon.Name.Location, name, parameterName);
                         }
-                        else
+                    }
+                    else
+                    {
+                        Debug.Assert(deconstructMethod is ErrorMethodSymbol);
+                    }
+                }
+                var boundSubpattern = new BoundSubpattern(
+                    subPattern,
+                    parameter,
+                    BindPattern(subPattern.Pattern, elementType, isError, diagnostics)
+                    );
+                patterns.Add(boundSubpattern);
+            }
+
+            return patterns.ToImmutableAndFree();
+        }
+
+        private bool ShouldUseITuple(
+            RecursivePatternSyntax node,
+            TypeSymbol declType,
+            DiagnosticBag diagnostics,
+            out NamedTypeSymbol iTupleType,
+            out MethodSymbol iTupleGetLength,
+            out MethodSymbol iTupleGetItem)
+        {
+            iTupleType = null;
+            iTupleGetLength = iTupleGetItem = null;
+            if (declType.IsTupleType)
+            {
+                return false;
+            }
+
+            if (node.Type != null)
+            {
+                // ITuple matching only applies if no type is given explicitly.
+                return false;
+            }
+
+            if (node.PropertyPatternClause != null)
+            {
+                // ITuple matching only applies if there is no property pattern part.
+                return false;
+            }
+
+            if (node.DeconstructionPatternClause == null)
+            {
+                // ITuple matching only applies if there is a deconstruction pattern part.
+                return false;
+            }
+
+            if (node.Designation != null)
+            {
+                // ITuple matching only applies if there is no designation (what type would the designation be?)
+                return false;
+            }
+
+            if (((CSharpParseOptions)node.SyntaxTree.Options).LanguageVersion < MessageID.IDS_FeatureRecursivePatterns.RequiredVersion())
```

Comment:
>LanguageVersion [](start = 62, length = 15)

Feels like we should use the LanguageVersion from the compilation (ie. max of all trees) rather than from a single tree.

Reply:
The compilation's language version isn't the max. It requires that the trees have the same language version (else it throws an exception) and uses that. However, I agree that Compilation.LanguageVersion should be used here.

---
In reply to: [206289340](https://github.com/dotnet/roslyn/pull/28859#discussion_r206289340) [](ancestors = 206289340)